examples/unit/naming/README.md

```diff
@@ -0,0 +1,120 @@
+
+```python
+    # Get the bucket & check it contains the region name somewhere in
+    # it (a pretty common naming convention).
+    bucket = stack.get_resource("rS3Bucket")
+    bucket.assert_property_value_matches_pattern("BucketName", r"^[a-z0-9-]*-xx-west-3[a-z0-9-]*$")
+
+    # The EFS volume in this template also has a name including substitutions,
+    # but this time the value is in a tag.
+    # For a pattern to match we will just check that is ends in "-vol".
+    efs_vol = stack.get_resource("rFileSystem")
+    # This last attribute is optional, and defaults to Tags which is used in many
+    # resources.
+    efs_vol.assert_tag_value_matches_pattern("Name", r"^[a-z0-9-]*-vol$", "FileSystemTags")
+
+```
+
+If a resource was incorrectly named and did not match the expected pattern, this would cause an assertion error to be raised like this:
+
+```
+E       AssertionError: Resource 'rFileSystem' tag 'Name' value 'my-test-xx-west-3-vol' did not match expected pattern '^[a-z]*-vol$'.
+```
+
+
+# Advanced Naming Convention Checks
+
+With the previous examples you can easily target individual resources to check their names meet expectations, but adding a test like this for every single resource in templates across your organisation is monotonous and time consuming. It is common for organisations to define conventions like all S3 buckets should have a naming convention like `{company name}-{AWS region}-{some name}`. This example covers the more advanced case for doing resource type to pattern conventions. This example is taken from the `test_naming_conventions` method of [test_naming_advanced.py](./test_naming_advanced.py).
+
+
+```python
+    # Ideally this dict would be coming from a common library
+    # function that you have shared between all your test cases
+    # (assuming consistency is the goal).
+    #
+    # This defines all the resource types we want to check, the pattern to match,
+    # and either the details of the Tag or Property that the name is held in.
+    type_patterns = {
+        "AWS::EFS::FileSystem": {
+            "Tag": "Name",
+            # This TagProperty is optional. The default is 'Tags',
+            # but some resources use a different property name for
+            # their tags.
+            "TagProperty": "FileSystemTags",
+            "Pattern": r"^[a-z0-9-]*-vol$",
+        },
+        "AWS::S3::Bucket": {
+            "Property": "BucketName",
+            "Pattern": r"^[a-z0-9-]*-xx-west-3[a-z0-9-]*$",
+        },
+        "AWS::S3::BucketPolicy": {
+            # The BucketPolicy type does not support custom names. If we do not
+            # want to set fail_on_missing_type=False when we call the assertion
+            # below then we need to include this type in the dict, and set it
+            # not to be checked.
+            # This approach ensures that types do not slip through unintentionally.
+            "Check": False
+        },
+    }
```

Comment:
Certainly might be the case where flexibility has been added in where it isn't required. At my company we are very much in the camp that anything that can have a name will have a name and a defined convention (we use CloudFormation Guard for ensuring the properties / tags are set - but it can't do this parameter substitution to do the convention enforcement). 

I can see the case where at least checking certain resources (like S3 buckets) contain the region name they exist in, but the resource level checks could be used for them without the stack level validation. 

From the general point of your review, it's this stack level validation which has the pain points - not the resource level changes? If so I can split that piece out into a separate PR if you want.